readthedocs/builds/models.py

```diff
@@ -448,6 +478,7 @@ class Build(models.Model):
     exit_code = models.IntegerField(_('Exit code'), null=True, blank=True)
     commit = models.CharField(
         _('Commit'), max_length=255, null=True, blank=True)
+    config = JSONField(_('Configuration used in the build'), default={})
```

Comment:
I like this, but I think this will increase _a lot_ the size of the db with not very useful data (because it will become old/inaccurate pretty soon). On the other hand, moving this to the cold storage I think that doesn't make sense since finally we won't be able to query these models and get the data we want.

Because of this problem, I was suggesting to save this config in the `Version` object and always save the config of the latest build per version. We don't really care what was the config of a build 2 months ago. This way, we will have _only one config JSON_ saved into our DB per (active) version and it will be always up to date.

Besides, answering the question: "how many project use `mkdocs`?" with the current approach is complicated since we will need to count _only_ the latest build per version. In the end, I think that in many cases what I'm suggesting is what we want here.

I want to make this stored data useful, help us to debug problems but also to make decisions when we need to know how users are using our own platform: "how much used is the `build/latest` image?" and more questions.

Reply:
Using the JSON example generated from the test and checking how many builds we do have per day, I can estimate that we will have an _increase of 1.5Mb_ **per day**.

```
In [13]: Build.objects.filter(date__lt=datetime.date(2018, 10, 15), date__gt=datetime.date(2018, 10, 14)).count()
Out[13]: 3278

In [14]: text_json = '{"version": "1", "formats": ["pdf"], "python": {"version": 3.5, "requirements": "requirements.txt", "install_with_pip": false, "install_with_setup": false, "extra_requi
    ...: rements": [], "use_system_site_packages": false}, "build": {"image": "readthedocs/build:2.0"}, "conda": null, "sphinx": {"builder": "sphinx", "configuration": null, "fail_on_warning
    ...: ": false}, "mkdocs": {"configuration": null, "fail_on_warning": false}, "doctype": "sphinx", "submodules": {"include": 1, "exclude": [], "recursive": true}}'
    ...: 

In [15]: len(text_json)
Out[15]: 505

In [16]: 505 * 3278 / 1024 ** 2
Out[16]: 1.5787029266357422
```

(please, verify my math and logic)

Reply:
1.5MB per day feels like not that much storage, to be honest. It's less about 500MB a year, give or take. Of course, doing this kind of stuff is how we keep filling up our DB, so probably best not to take the naive approach.

I do see value in having recent changes of the config files stored. I wonder if there is a good way to store it on the Version, but keep a historical copy of all changes. We don't want to keep 500 copies of the same exact data structure, but being able to diff them over time feels very useful, especially when a build goes from passing to failing, or similar.

Reply:
That might be overkill, perhaps we could just only save the config when it differs from the previously found one on the `Version`? Something that limits data so we're not just storing the same data over and over.

Reply:
I'm not sure if there is an easy way, I'd like to have this to show to users the configuration that was used on each build, if we don't save the full config we need to query the previous one before saving (this is easy) to see if they differ. If we want to show the config for a specific build, we need to know the previous one. So, maybe another model? Or maybe a foreign key to the previous `Build`?

Reply:
I _personally_ prefer to take an action about this _before_ merging.

@ericholscher eric I like your proposal. I think it's "simple enough" to implement and shouldn't complicate things too much if we save "the whole config" if "we detect that differs", instead of saving a diff that later will be more complicated to generate the real config JSON text.

This way, we can query "what's the latest build of this project where the `config` attribute is not `None`" and that will be the `config` used for the current Build (and we can show it to the user as @stsewd suggested).

Reply:
So, `Build` has a date field, we can use that to query the previous build with a _valid_ config